src/workerd/api/sockets.c++

```diff
@@ -119,4 +100,21 @@ jsg::Promise<void> Socket::close(jsg::Lock& js) {
   }, [this](jsg::Lock& js, jsg::Value err) { return errorHandler(js, kj::mv(err)); });
 }
 
+jsg::Promise<void> Socket::handleProxyStatus(
+    jsg::Lock& js, kj::Promise<kj::HttpClient::ConnectRequest::Status> status) {
+  auto& context = IoContext::current();
+  auto result = context.awaitIo(js, kj::mv(status),
+      [this](jsg::Lock& js, kj::HttpClient::ConnectRequest::Status&& status) -> void {
```

Comment:
Since you are capturing [this] here, you need to ensure the object cannot be destroyed before the promise resolves.

Typically this is ensured by returning the promise to JavaScript. The JSG glue for methods that return promises ensures that a strong ref to the object gets attached to the promise. However, this promise is not being returned to JS, it's just being stored in a member variable. This doesn't create any such attachment.

The best way to solve this is probably to capture `[self = JSG_THIS]` instead of `[this]`.

You also should NOT store the resulting promise in a member of `Socket`. Since it's a JSG promise, unlike KJ, you don't need to store a reference to prevent it from being canceled -- it'll run to completion either way. But if you store a reference in `Socket`, and the continuation callback also captures a strong reference back to `Socket`, this might create a cycle. You could solve the cycle by making sure all objects are GC-visitable but it's probably easier to just not store the promise in a member since it's not useful anyway.